File: bdp-devops/ambari/ambari-server/src/main/java/org/apache/ambari/server/utils/ModuleVersion.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.apache.ambari.server.utils;


import java.util.regex.Matcher;
import java.util.regex.Pattern;

import org.apache.commons.lang.StringUtils;

/**
 * This class should be used to compare module(service) versions.
 * Base method which should be used is parse(..), This method will validate and parse
 * version which you will pass as parameter, and return object of current class with
 * parsed version. Same thing you should do with another version, with which you are
 * planning to compare previous one. After that, use method compare to get final result.
 */

public class ModuleVersion implements Comparable<ModuleVersion> {

  // RE for different version formats like N.N.N.N-bN, N.N.N-hN-bN
  private static final String VERSION_WITH_HOTFIX_AND_BUILD_PATTERN = "^([0-9]+).([0-9]+).([0-9]+).([0-9]+)-h([0-9]+)-b([0-9]+)";
  private static final String VERSION_WITH_BUILD_PATTERN = "^([0-9]+).([0-9]+).([0-9]+).([0-9]+)-b([0-9]+)";

  // Patterns for previous RE
  private static final Pattern PATTERN_WITH_HOTFIX = Pattern.compile(VERSION_WITH_HOTFIX_AND_BUILD_PATTERN);
  private static final Pattern PATTERN_WITHOUT_HOTFIX = Pattern.compile(VERSION_WITH_BUILD_PATTERN);

  // Parts of version
  private int apacheMajor;
  private int apacheMinor;
  private int internalMinor;
  private int internalMaint;
  private int hotfix;
  private int build;


  public ModuleVersion(int apacheMajor, int apacheMinor, int internalMinor, int internalMaint, int hotfix, int build) {
    this.apacheMajor = apacheMajor;
    this.apacheMinor = apacheMinor;
    this.internalMinor = internalMinor;
    this.internalMaint = internalMaint;
    this.hotfix = hotfix;
    this.build = build;
  }

  /**
   * Method which will parse module version
   * which user passed as parameter. Also
   * in this method version will be validated.
   * @param moduleVersion string
   * @return MpackVersion instance which contains parsed version
   * */
  public static ModuleVersion parse(String moduleVersion) {
    Matcher versionMatcher = validateModuleVersion(moduleVersion);
    ModuleVersion result = null;

    if (versionMatcher.pattern().pattern().equals(VERSION_WITH_HOTFIX_AND_BUILD_PATTERN)) {
      result = new ModuleVersion(Integer.parseInt(versionMatcher.group(1)), Integer.parseInt(versionMatcher.group(2)),
              Integer.parseInt(versionMatcher.group(3)), Integer.parseInt(versionMatcher.group(4)),
              Integer.parseInt(versionMatcher.group(5)), Integer.parseInt(versionMatcher.group(6)));

    } else {
      result = new ModuleVersion(Integer.parseInt(versionMatcher.group(1)), Integer.parseInt(versionMatcher.group(2)),
              Integer.parseInt(versionMatcher.group(3)), Integer.parseInt(versionMatcher.group(4)), 0,
              Integer.parseInt(versionMatcher.group(5)));

    }

    return result;
  }

  /**
   * Method validate module version not to be
   * empty or null. Also check if passed version
   * has valid format.
   * @param version string
   * @return Matcher for passed version
   * @throws IllegalArgumentException() if version empty/null/not valid
   */
  private static Matcher validateModuleVersion(String version) {
    if (StringUtils.isEmpty(version)) {
      throw new IllegalArgumentException("Module version can't be empty or null");
    }

    String moduleVersion = StringUtils.trim(version);

    Matcher versionMatcher = PATTERN_WITH_HOTFIX.matcher(moduleVersion);
    if (!versionMatcher.find()) {
      versionMatcher = PATTERN_WITHOUT_HOTFIX.matcher(moduleVersion);
      if (!versionMatcher.find()) {
        throw new IllegalArgumentException("Wrong format for module version, should be N.N.N.N-bN or N.N.N-hN-bN");
      }
    }

    return versionMatcher;
  }

  @Override
  public int compareTo(ModuleVersion other) {
    int result = this.apacheMajor - other.apacheMajor;
    if(result == 0) {
      result = this.apacheMinor - other.apacheMinor;
      if(result == 0) {
        result = this.internalMinor - other.internalMinor;
        if(result == 0) {
          result = this.internalMaint - other.internalMaint;
          if(result == 0) {
            result = this.hotfix - other.hotfix;
            if(result == 0) {
              result = this.build - other.build;
            }
          }
        }
      }
    }
    return result > 0 ? 1 : result < 0 ? -1 : 0;
  }

  @Override
  public boolean equals(Object o) {
    if (this == o) return true;
    if (o == null || getClass() != o.getClass()) return false;

    ModuleVersion that = (ModuleVersion) o;

    if (apacheMajor != that.apacheMajor) return false;
    if (apacheMinor != that.apacheMinor) return false;
    if (build != that.build) return false;
    if (hotfix != that.hotfix) return false;
    if (internalMaint != that.internalMaint) return false;
    if (internalMinor != that.internalMinor) return false;

    return true;
  }

  @Override
  public int hashCode() {
    int result = apacheMajor;
    result = 31 * result + apacheMinor;
    result = 31 * result + internalMinor;
    result = 31 * result + internalMaint;
    result = 31 * result + hotfix;
    result = 31 * result + build;
    return result;
  }
}

File: bdp-devops/ambari-versions/ambari-3.0.0.0-SNAPSHOT/ambari-common/src/main/python/resource_management/libraries/functions/module_version.py
"""
  Licensed to the Apache Software Foundation (ASF) under one
  or more contributor license agreements.  See the NOTICE file
  distributed with this work for additional information
  regarding copyright ownership.  The ASF licenses this file
  to you under the Apache License, Version 2.0 (the
  "License"); you may not use this file except in compliance
  with the License.  You may obtain a copy of the License at

      http://www.apache.org/licenses/LICENSE-2.0

  Unless required by applicable law or agreed to in writing, software
  distributed under the License is distributed on an "AS IS" BASIS,
  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  See the License for the specific language governing permissions and
  limitations under the License.
"""

import re

"""
 This class should be used to compare module(service) versions.
 Base method which should be used is parse(..), This method will validate and parse
 version which you will pass as parameter, and return object of current class with
 parsed version. Same thing you should do with another version, with which you are
 planning to compare previous one. After that, use "==", "<", ">" to get final result.
"""
class ModuleVersion(object):
  __module_version_pattern = "(?P<aMajor>[0-9]+).(?P<aMinor>[0-9]+).(?P<iMinor>[0-9]+).(?P<iMaint>[0-9]+)(-h(?P<hotfix>[0-9]+))*-b(?P<build>[0-9]+)"
  __module_version_regex = re.compile(__module_version_pattern)

  def __init__(self, apache_major, apache_minor, internal_minor, internal_maint, hotfix, build):
    """
    :type apache_major int
    :type apache_minor int
    :type internal_maint int
    :type internal_minor int
    :type hotfix int
    :type build int
    """
    self.__apache_major = int(apache_major)
    self.__apache_minor = int(apache_minor)
    self.__internal_maint = int(internal_maint)
    self.__internal_minor = int(internal_minor)
    self.__hotfix = int(hotfix) if hotfix else 0  # hotfix is optional group
    self.__build = int(build)

  def __repr__(self):
    return "{0}.{1}.{2}.{3}-h{4}-b{5}".format(*self.to_list())

  def to_list(self):
    """
    Return version elements as list

    :rtype list
    """
    return [
      self.__apache_major,
      self.__apache_minor,
      self.__internal_minor,
      self.__internal_maint,
      self.__hotfix,
      self.__build
    ]

  def __cmp__(self, other):
    """
    :type other ModuleVersion

    :raise TypeError
    """
    if other and not isinstance(other, self.__class__):
      raise TypeError("Operand type is different from {0}".format(self.__class__.__name__))

    r = 0
    x = self.to_list()
    y = other.to_list()

    for i in range(0, len(x)):
      r = x[i] - y[i]
      if r != 0:
        break

    return 1 if r > 0 else -1 if r < 0 else 0

  @classmethod
  def parse(cls, module_version):
    """
      Parse string to module version

      :type module_version str
      :rtype ModuleVersion
      """
    matcher = cls.validate(module_version)
    return ModuleVersion(
      matcher.group("aMajor"),
      matcher.group("aMinor"),
      matcher.group("iMinor"),
      matcher.group("iMaint"),
      matcher.group("hotfix"),
      matcher.group("build")
    )

  @classmethod
  def validate(cls, module_version):
    """
    Check if provided version is valid. If version is valid will return match object
    or will raise exception.

    :param module_version version to check
    :type module_version str

    :rtype __Match[T] | None

    :raise ValueError
    """

    if not module_version:
      raise ValueError("Module version can't be empty or null")

    version = module_version.strip()

    if not version:
      raise ValueError("Module version can't be empty or null")

    matcher = cls.__module_version_regex.match(version)

    if not matcher:
      raise ValueError("{0} is not a valid {1}".format(version, cls.__name__))

    return matcher

  @property
  def apache_major(self):
    return self.__apache_major

  @property
  def apache_minor(self):
    return self.__apache_minor

  @property
  def internal_minor(self):
    return self.__internal_minor

  @property
  def internal_maint(self):
    return self.__internal_maint

  @property
  def hotfix(self):
    return self.__hotfix

  @property
  def build(self):
    return self.__build



